src/Lib.P2P/Cid.cs

```diff
@@ -0,0 +1,610 @@
+{
+    /// <summary>
+    ///  Identifies some content, e.g. a Content ID.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///   A Cid is a self-describing content-addressed identifier for distributed systems.
+    ///   </para>
+    ///   <para>
+    ///   Initially, IPFS used a <see cref="MultiHash"/> as the CID and this is still supported as <see cref="Version"/> 0.
+    ///   Version 1 adds a self describing structure to the multi-hash, see the <see href="https://github.com/ipld/cid">spec</see>. 
+    ///   </para>
+    ///   <note>
+    ///   The <see cref="MultiHash.Algorithm">hashing algorithm</see> must be "sha2-256" for a version 0 CID.
+    ///   </note>
+    /// </remarks>
+    /// <seealso href="https://github.com/ipld/cid"/>
+    [JsonConverter(typeof(CidJsonConverter))]
+    public class Cid : IEquatable<Cid>
+    {
+        /// <summary>
+        ///   The default <see cref="ContentType"/>.
+        /// </summary>
+        public const string DefaultContentType = "dag-pb";
+
+        private string encodedValue;
+        private int version;
+        private string encoding = MultiBase.DefaultAlgorithmName;
+        private string contentType = DefaultContentType;
+        private MultiHash hash;
+
+        /// <summary>
+        ///   Throws if a property cannot be set.
+        /// </summary>
+        /// <exception cref="NotSupportedException">
+        ///   When a property cannot be set.
+        /// </exception>
+        /// <remarks>
+        ///   Once <see cref="Encode"/> is invoked, the CID's properties
+        ///   cannot be set.
+        /// </remarks>
+        private void EnsureMutable()
+        {
+            if (encodedValue != null) throw new NotSupportedException("CID cannot be changed.");
+        }
+
+        /// <summary>
+        ///   The version of the CID.
+        /// </summary>
+        /// <value>
+        ///   0 or 1.
+        /// </value>
+        /// <remarks>
+        ///   <para>
+        ///   When the <see cref="Version"/> is 0 and the following properties
+        ///   are not matched, then the version is upgraded to version 1 when any 
+        ///   of the properties is set.
+        ///   <list type="bullet">
+        ///   <item><description><see cref="ContentType"/> equals "dag-pb"</description></item>
+        ///   <item><description><see cref="Encoding"/> equals "base58btc"</description></item>
+        ///   <item><description><see cref="Hash"/> algorithm name equals "sha2-256"</description></item>
+        ///   </list>
+        ///   </para>
+        ///   <para>
+        ///   </para>
+        ///   The default <see cref="Encoding"/> is "base32" when the
+        ///   <see cref="Version"/> is not zero.
+        /// </remarks>
+        public int Version
+        {
+            get => version;
+            set
+            {
+                EnsureMutable();
+                if (version == 0 && value > 0 && Encoding == "base58btc") encoding = "base32";
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)